src/shell/shell_cmds.cpp

```diff
@@ -238,18 +236,53 @@ void DOS_Shell::CMD_DELETE(char * args) {
 
 void DOS_Shell::CMD_HELP(char * args){
 	HELP("HELP");
-	bool optall=ScanCMDBool(args,"ALL");
+	const bool optall = ScanCMDBool(args, "ALL");
 	/* Print the help */
-	if (!optall) WriteOut(MSG_Get("SHELL_CMD_HELP"));
-	Bit32u cmd_index=0,write_count=0;
+	if (!optall && !*args)
+		WriteOut(MSG_Get("SHELL_CMD_HELP"));
+	upcase(args);
+	uint32_t cmd_index = 0, write_count = 0;
+	bool show = false;
 	while (cmd_list[cmd_index].name) {
-		if (optall || !cmd_list[cmd_index].flags) {
-			WriteOut("<\033[34;1m%-8s\033[0m> %s",cmd_list[cmd_index].name,MSG_Get(cmd_list[cmd_index].help));
-			if (!(++write_count % 24))
-				CMD_PAUSE(empty_string);
+		bool match_cmd = *args && !strcmp(args, cmd_list[cmd_index].name);
+		bool all_cmds = !*args && !cmd_list[cmd_index].flags;
+		if (optall || match_cmd || all_cmds) {
+			if (match_cmd && !optall) {
```

Comment:
Thanks @Wengier ; compared to the previous block of logic:

``` c++
	while (cmd_list[cmd_index].name) {
		if (optall || (*args && !strcmp(args, cmd_list[cmd_index].name)) ||
		    (!*args && !cmd_list[cmd_index].flags)) {
			if (*args && !strcmp(args, cmd_list[cmd_index].name) &&
			    !optall)
```

I would say this is an improvement; thanks!

As someone who hasn't seen this code before, there's still a lot to mentally juggle with all these conditionals (and all the permutations of their states) - and that's without a clear understanding of what it's trying to accomplish.

I realize there might not be more simplifications though; so will keep looking at this and try to make more concrete suggestions if anything else comes to mind.

Reply:
I see, so I guess the code was still not so straightforward to understand. I have updated the source code again and added some comments, which hopefully make it easier to understand. Please check it out.

Reply:
The comments help my understanding a lot @Wengier , thank you!